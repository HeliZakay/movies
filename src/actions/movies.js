import uuid from "uuid";
import database from "../firebase/firebase";
import moment from "moment";



// ADD_MOVIE
export const addMovie = (movie) =>  ({
      type: 'ADD_MOVIE',
      movie
});
  
export const startAddMovie = (movieData = {}) => {
  return (dispatch, getState) => {
    const {
      movieName="",
      personName="",
      score=0,
      content="",
      createdAt="0",
      genres     
     } = movieData;
     let movie = {
      movieName
    };
    if (genres) {
      movie = {
        ...movie,
        genres
      }
    }
    const review = {
      personName,
      score,
      content,
      createdAt: moment(createdAt).format(), 
      userUid: getState().auth.uid
    }
     return database.ref("movies").push(movie).then( (ref) => {
        return database.ref(`movies/${ref.key}/reviews`).push(review).then((r) => {
          dispatch(addMovie({
            id: ref.key,
            movieName,
            reviews: [{...review, id: r.key}],
            genres
          }));
        });
     });
  };
};


// export const  = () => {
//   return (dispatch, getState) => {
   
//     return database.ref("movies").once("value").then((snapshot) => {
//       snapshot.forEach((childSnapshot) => {
//         const movieName = childSnapshot.val().movieName;
//         const id = childSnapshot.val().id;
//         const genres = getMovieGenres(movieName);
//         if (genres) {
//           return database.ref(`movies/${id}`).update({genres: genres});
//         }
//       });
//     });
//   }
// }

export const startAddReview = ({movieId, score, personName, content, createdAt}) => {
  return (dispatch, getState) => {
     const reviewItem = {
      personName,
      score,
      content,
      createdAt: moment(createdAt).format(), 
      userUid: getState().auth.uid
    };
     return database.ref(`movies/${movieId}/reviews`).push(reviewItem).then( (ref) => {
        return dispatch(addReview(movieId, {
          id: ref.key,
          ...reviewItem
        }));
     });
  };
};

export const addReview = (movieId, review) => ({
  type: "ADD_REVIEW",
  review,
  movieId
});


// EDIT_MOVIE
export const editMovie = ({movieId, reviewId, updatedReview}) => ({
  type: 'EDIT_MOVIE',
  movieId,
  reviewId,
  updatedReview
});

export const startEditMovie = ({movieId, reviewId, updatedReview}) => {
  return (dispatch) => {
     return database.ref(`movies/${movieId}/reviews/${reviewId}`).update(updatedReview).then(() => {
        dispatch(editMovie({movieId, reviewId, updatedReview}));
     });
  };
};

 // REMOVE_MOVIE
 export const removeMovie = ({movieId, reviewId} = {}) => ({
  type: 'REMOVE_MOVIE',
  movieId,
  reviewId
});

export const removeMovieCompletely = ({movieId}) => ({
  type: "REMOVE_MOVIE_COMPLETELY",
  movieId
});

export const startRemoveMovie = ({movieId, reviewId} = {}) => {
  return (dispatch) => {
     return database.ref(`movies/${movieId}/reviews/${reviewId}`).remove().then( () => {
        return database.ref(`movies/${movieId}`).once("value").then((snapshot) => {
          if (!snapshot.val().reviews) {
            return database.ref(`movies/${movieId}`).remove().then(() => {
              dispatch(removeMovieCompletely({movieId}));
            });
          } else {
            dispatch(removeMovie({movieId, reviewId}));
          }
        });
     });
  };
};

 

  //SET_MOVIES
  export const setMovies = (movies) => ({
    type: "SET_MOVIES",
    movies
  });


  export const startSetMovies = () => {
    return (dispatch) => {
      return database.ref("movies").once("value").then( (snapshot) => {
        const movies = [];
        snapshot.forEach((movieSnapshot) => {
            const reviews = [];
            if (movieSnapshot.val().reviews){
              const reviewsObj = movieSnapshot.val().reviews;
              for (const [key, value] of Object.entries(reviewsObj)) {
                reviews.push({
                  id: key,
                  ...value
                });
              }
            }
              movies.push({
                 id: movieSnapshot.key,
                 ...movieSnapshot.val(),
                 reviews
              });
        });
        dispatch(setMovies(movies));
      }); 
    };
  };

export const unifyReviews = () => {
  database.ref("movies/-MGnz68SXXFy0sVqeoP3").remove();
    
}