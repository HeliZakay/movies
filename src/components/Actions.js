import React from "react";
import { Link} from 'react-router-dom';

export default () => {
    return (
        <div className="actions-section">
        <div className="content-container">
        <div className="actions-section__wrapper">
        
        <div className="actions-section__item">
            <Link className="actions-section__link" to="/create">
            <div className="actions-section__background">
            <img className="actions-section__img" src="images/popcorn.png"></img> 
            </div>
            <h3 className="actions-section__text">Add a movie</h3>
            
            </Link>
        </div>
        <div className="actions-section__item">
            <Link className="actions-section__link" to="/friends">
            <div className="actions-section__background">
            <img className="actions-section__img" src="images/familyguy.png"></img>  
            </div>
            <h3 className="actions-section__text">Add a friend</h3>
           
            </Link>
        </div>
        <div className="actions-section__item">
            <Link className="actions-section__link" to="/contact">
            <div className="actions-section__background">
            <img className="actions-section__img" src="images/envelop.png"></img>  
            </div>
            <h3 className="actions-section__text">Send feedback</h3>
            
            </Link>
        </div>
        </div>  
        </div>          
        </div>
    );
};



