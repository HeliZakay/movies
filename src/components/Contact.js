import React from "react";
import {connect} from "react-redux";

export class Contact extends React.Component {
    onSubmit = (event) => {
        event.preventDefault();
        this.props.history.push("/homepage");
    }
    render() {
        return (
            <form
            action="/" 
            method="post" 
            encType="text/plain"
            // onSubmit={this.onSubmit}
            className="form" >

            <div className="page-header">
            <div className="content-container--form">
                 <h2 className="page-header__title">Send Feedback</h2>
            </div>
        </div>
        <div className="content-container--form">
           <p>
            The goal here is to make your experience as much pleasant as possible!
           </p>
           <textarea className="textarea text-area--contact"
                        placeholder="Write your feedback here"
                        name="feedback" autoFocus>
        </textarea>
        <div>
                <button type="submit" className="button button--form"><strong>Send!</strong></button>
        </div>
        </div>
        </form>

        );
    };
}

    
 export default connect()(Contact);
  
