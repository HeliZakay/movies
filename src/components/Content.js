import React from 'react';
import {connect} from "react-redux";

export  class Content extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            readMore: false
        }
    }
    readMore = () => {
        this.setState({readMore: !this.state.readMore});
    }
    shortenContent = (content) => {
        if(content.length< 150) {
            return <p>"{content}"</p>;
        }
        else {
            return <div>
            {!this.state.readMore ?
            <p>
            "{content.slice(0,150)}..."
            <a onClick={this.readMore}>
            {this.props.language === "English"?  "Read More" : "קרא עוד "}
            </a>
            </p>:
            <p>
            "{content}"
            <a onClick={this.readMore}>
            {this.props.language === "English"?  "Read Less" : "קרא פחות "}
            </a>
            </p>
            }
            </div>;
        }
    }
    render() {
        return (
            <div className="content">
             {this.props.content && <div className="card-text"> {this.shortenContent(this.props.content)}</div>}           
            </div>
        );
    }
};



const mapStateToProps = (state) => ({
    language: state.auth.language
});

export default connect(mapStateToProps)(Content);
   