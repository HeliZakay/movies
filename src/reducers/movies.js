// Movie Reducer

export default (state = [], action) => {
    switch (action.type) {
      case 'ADD_MOVIE':
        return [
          ...state,
          action.movie
        ];
        case 'EDIT_MOVIE':
          return state.map((movie) => {
            if (movie.id === action.id) {
              return {
                ...movie,
                ...action.updates
              };
            } else {
              return movie;
            };
          });
      case "ADD_MOVIE_TO_WATCH_LIST":
        return state.map((movie) => {
          if(movie.id === action.id) {
            return {
              ...movie,
              watchList: true
            };
          } else {
            return movie;
          }
        });
      case "REMOVE_MOVIE_FROM_WATCH_LIST":
        return state.map((movie) => {
          if(movie.id === action.id) {
            return {
              ...movie,
              watchList: false
            };
          } else {
            return movie;
          }
        });
      case 'REMOVE_MOVIE':
        return state.filter(({ id }) => id !== action.id);
      default:
        return state;
    }
  };
