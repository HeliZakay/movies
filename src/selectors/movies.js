import FriendDialog from "../components/FriendDialog";

  export const getHname = (movies, movieName) => {
    const movie = movies.filter((movie) => {
      return movie.movieName === movieName;
    });
    if (movie && movie[0] && movie[0].hname) {
      return movie[0].hname;
    }
    else {
      return undefined;
    }
  }
  
  
  export const didIGaveStarToReview = (stars, uid) => {
    let result = false;
    if (stars) {
      stars.forEach((star) => {
        if (star.uid === uid) {
          result = true;
        }
      });
    }
    return result;
  }
  
  export const sortReviewsByDateAndPerson = (reviews, person) => { 
    if (!person) {
      return reviews.sort((a,b) => { 
        return a.createdAt < b.createdAt? 1:-1;
      });
    } else {
      const sortedByDate = reviews.sort((a,b) => { 
        return a.createdAt < b.createdAt? 1:-1;
      });
     const personReview =  reviews.filter((review) => {
        return review.personName.toLowerCase().includes(person.toLowerCase());
      });
      return [...personReview, ...sortedByDate.filter((review)=> {
        return !review.personName.toLowerCase().includes(person.toLowerCase());
      })];
    }   
  };

  
const computeAverageScore = (movie) => {
  let sum = 0;
  movie.reviews.forEach((review) => {
      sum += Number(review.score);
  });
  const result = sum/movie.reviews.length;
  return  Math.round(result*2)/2;;
}

// Get visible movies
export default (movies, { text, sortBy, person, genres}, language) => {
    return movies.filter((movie) => {
      let textMatch;
      if (language === "English") {
        textMatch = movie.movieName.toLowerCase().includes(text.toLowerCase());
      } else {
        textMatch = String(movie.hname).includes(text);
      }
      let personMatch = false;
      movie.reviews.forEach((review) => {
        if (review.personName.toLowerCase().includes(person.toLowerCase())) {
          personMatch = true;
        }
      });
      let genreMatch = false;
      if (!genres) {
        genreMatch = true;
      } else if (genres.length ===0) {
        genreMatch = true;
      } else if (!movie.imdbData.Genre) {
        genreMatch = false;
      } else {
        const genresArray = movie.imdbData.Genre.split(", ");
        genresArray.forEach((genre) => {
          if (genres.includes(genre)) {
            genreMatch = true;
          }
        });
      }
      return genreMatch && textMatch && personMatch;
    }).map((movie) => {
      const sorted = sortReviewsByDateAndPerson(movie.reviews, person);
      return {...movie, reviews: sorted};
    }).sort((a, b) => {
      if (sortBy === 'date') {
        return a.reviews[0].createdAt < b.reviews[0].createdAt ? 1 : -1;
      } else if (sortBy === 'score') {
        return computeAverageScore(a) < computeAverageScore(b) ? 1 : -1;
      }
    });
  };

  export const getMovieById = (movies, movieId) => {
    let result = undefined;
    movies.forEach((movie) => {
      if(movie.id === movieId) {
        result = movie;
      }
    });
    return result;
  };

  export const getMovieReviews = (movieId, movies) => {
    let reviews = [];
    movies.forEach((movie) => {
      if (movie.id === movieId) {
        reviews=movie.reviews;
      }
    });
    return reviews;
  }

  export const isMovieNameExistsAlready = (movieName, movies) => {
    let result = undefined;
    movies.forEach((movie) => {
      if (movie.movieName === movieName) {
        result=movie.id;
      }
    });
    return result;
  }
export const findMovie = (movies, movieName, personName, uid, type) => {
  const movieArr =  movies.filter((movie) => {
    return movie.movieName === movieName;
  });  
  if(movieArr && movieArr[0]) {
    const movie = movieArr[0];
    const oldReviews = movie.reviews;
    const filteredReviews = oldReviews.filter((review) => {
      if(type==="newReview") {
        return review.personName !== personName;
      }
      else {
        return review.userUid !== uid;
      }
    });
    const firstReview = oldReviews.filter((review) => {
      if(type==="newReview") {
        return review.personName === personName;
      }
      else {
        return review.userUid === uid;
      }
    });
    const sortedReviews= [...firstReview, ...filteredReviews];
    return {
      ...movie,
      reviews: sortedReviews
    };
  } else {
    return undefined;
  }
  
}

export const getUnseenMovies = (movies, uid, watchlist) => {
  return movies.filter((movie)=> {
    const reviewsUids = movie.reviews.map((review)=> {
      return review.userUid;
    });
    return (!reviewsUids.includes(uid) && !watchlist.includes(movie.id) && computeAverageScore(movie) >=7);
  });
}

export const takeTop = (array, num) => {
  if (array.length <=num) {
    return array;
  }
  else {
    return array.slice(0,num);
  }
}

export const didFriendReviewedMovie = (reviews, friendId) => {
  
  const result = reviews.filter((review) => {
    return review.userUid === friendId;
  });
  return result.length > 0;

}

const didSendReviewToMovie = ({reviews, uid}) => {
  let reviewId;
  reviews.forEach((review) => {
    if (review.userUid === uid) {
      reviewId = review.id;
    }
  });
  return reviewId;
}

export const getMyMovies = (movies, uid) => {
  return movies.filter((movie) => {
    const myReviewId = didSendReviewToMovie({reviews: movie.reviews, uid});
    if (myReviewId) {
      let hightRatedByMe = false;
      movie.reviews.forEach((review) => {
        if(review.id === myReviewId && review.score >=8){
          hightRatedByMe = true;
        }
      });
      return hightRatedByMe;
    } else {
      return false;
    }
  });
} 