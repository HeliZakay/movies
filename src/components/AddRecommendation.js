import React from "react";
import RecommendationForm from "./RecommendationForm";
import {connect} from "react-redux";
import {startAddMovie} from "../actions/movies";

export class AddRecommendation extends React.Component {
   onSubmit = (movie) => {
      this.props.startAddMovie(movie);
      this.props.history.push("/homePage");
   };
   render() {
      return (
         <div>
         <h1>Add Recommendation</h1>
         <RecommendationForm
           onSubmit={this.onSubmit}
          />
         </div>
      );
   }
}


const mapDispatchToProps = (dispatch) => ({
   startAddMovie: (movie) => dispatch(startAddMovie(movie))
});
   
export default connect(undefined, mapDispatchToProps)(AddRecommendation);
