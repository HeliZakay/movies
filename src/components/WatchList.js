import React from "react";
import { connect } from "react-redux";
import MovieCard from "./MovieCard";
import getVisibleMovies from "../selectors/movies";
import {getWatchListMovies} from "../selectors/watchList";
import MoviesListFilters from "./MoviesListFilters";


export const WatchList = (props) => {
    return (
        <div>
        <div className="page-header">
        <div className="content-container">
        <h2 className="page-header__title">
        {props.language === "English"? "My Watch List": "רשימת הצפייה שלי"}
        </h2>
        </div>
         
        </div>
        <div className="content-container">
        <MoviesListFilters />
            {props.movies.length ===0 ?
            (
                
                <p className="friends__error"><em>Add movies to your watch list from homepage!</em></p>
               
            ) : (
                <div className="row">
               
                {props.movies.map((movie) => {
                    return (
                        <div 
                        key={movie.id} 
                        className=" col-sm-12 col-md-6 col-lg-4">
                        <MovieCard  {...movie} /> 
                        </div>
                    );
                })}
                </div>
            )}
        
            
           
        </div>
        </div>
    );
};

const mapStateToProps = (state) => {
    return {
        movies: getVisibleMovies(getWatchListMovies(state.movies, state.watchList), state.filters),
        language: state.auth.language
    };
};
export default connect(mapStateToProps)(WatchList);


