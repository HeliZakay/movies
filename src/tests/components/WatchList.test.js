import React from "react";
import {shallow} from "enzyme";
import {WatchList} from "../../components/WatchList";
import movies from "../fixtures/movies";

test("should render MoviesList with movies", () => {
    const wrapper = shallow(<WatchList movies={movies} />);
    expect(wrapper).toMatchSnapshot();
});

test("should not render MoviesList if there are no movies", () => {
    const wrapper = shallow(<WatchList movies={[]} />);
    expect(wrapper).toMatchSnapshot();
});