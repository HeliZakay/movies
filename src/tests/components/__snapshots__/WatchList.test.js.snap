// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should not render MoviesList if there are no movies 1`] = `
<div>
  <h2>
    My Watching List
  </h2>
  <[object Object] />
  <p>
    No movies yet- Add movies to your watching list from homepage
  </p>
</div>
`;

exports[`should render MoviesList with movies 1`] = `
<div>
  <h2>
    My Watching List
  </h2>
  <[object Object] />
  <[object Object]
    content="Cute"
    createdAt={"1970-01-01T00:00:04.000Z"}
    id="1"
    key="1"
    movieName="Tarazan got something"
    personName="Dorit"
    score={8}
    watchList={false}
  />
  <[object Object]
    content="the movie is OK"
    createdAt={"1970-01-01T00:00:01.000Z"}
    id="2"
    key="2"
    movieName="You've got mail"
    personName="Michal"
    score={10}
    watchList={false}
  />
  <[object Object]
    content="Very unique"
    createdAt={"1970-01-01T00:00:02.000Z"}
    id="3"
    key="3"
    movieName="Persian Prince"
    personName="Heli"
    score={7}
    watchList={true}
  />
</div>
`;
